File: Javascript/3.Ottobre/53.DecimalToBinary-02Ottobre.js
//todo Dato un numero intero non negativo, restituisci la sua rappresentazione binaria come stringa.
 
//! Un numero binario utilizza solo le cifre 0 e 1 per rappresentare qualsiasi numero. 
//! Per convertire un numero decimale in binario, dividi ripetutamente il numero per 2 e registra il resto. Ripeti finché il numero non è zero. Leggi i resti registrati per ultimi fino al primo. Ad esempio, per convertire 12 in binario:

//! 12 ÷ 2 = 6 resto 0
//! 6 ÷ 2 = 3 resto 0
//! 3 ÷ 2 = 1 resto 1
//! 1 ÷ 2 = 0 resto 1
//! 12 in binario è 1100.


//? Tempo di esecuzione:
//* Non completato 

//?Codice scritto senza utilizzo di ChatGpt

function toBinary(decimal) {

  let tot = [];
  let number = decimal

  for(let i = 0; i < number; i++) {
    let rest = number % 2;
    tot.push(rest)
    number = number / 2
  }

  return tot;
}

console.log(toBinary(5))

//?Codice scritto da ChatGpt

function toBinary(decimal) {
  if (decimal === 0) return "0"; //! OK

  let result = ""; //! Fa tornare direttamente una stringa

  while (decimal > 0) {
    result = (decimal % 2) + result;
    decimal = Math.floor(decimal / 2);
  }

  return result;
}

console.log(toBinary(5)); // "101"

//* ✅ PRO

//* Hai capito il concetto di resto / 2

//* Usi variabili chiare (rest, number)

//* La logica di base è corretta

//* Buon ragionamento algoritmico 💪

//! ❌ CONTRO

//! for sbagliato: number cambia → ciclo imprevedibile

//! Manca Math.floor() → numeri decimali

//! I bit sono al contrario (serve reverse)

//! Ritorni un array, non una stringa

//! Il ciclo può girare più del necessario

//? Versione Chat Gpt 

//todo Considerazioni: 

//! N.B.

File: Python/3.Ottobre/53.Decimal_to_binary-02Ottobre.py
# =========================================================
# FILE TEMPLATE - ANALISI ESERCIZI PYTHON
# Incolla direttamente questo file nel progetto VS Code
# =========================================================

# //Prompt ChatGpt: questo è il modello. Invierò consegna esercizio e codice, e vorrei feedback strutturato

# =========================================================
# //todo Consegna dell'esercizio
# =========================================================

# Dato un numero intero non negativo, restituisci la sua
# rappresentazione binaria come stringa.

# =========================================================
# //* Codice scritto senza utilizzo di ChatGpt
# =========================================================

def to_binary(decimal):
    number_converted = []
    divide_number = decimal

    while(divide_number > 0):
        if divide_number % 2 == 0:
            number_converted.insert(0, "0")
        else:
            number_converted.insert(0, "1")

        divide_number = divide_number // 2

    return "".join(number_converted)

print(to_binary(12))

# =========================================================
# //* Codice scritto da ChatGpt (versione migliorata)
# =========================================================

def to_binary(decimal):
    if decimal == 0:
        return "0"

    binary = []

    while decimal > 0:
        binary.append(str(decimal % 2))
        decimal //= 2

    return "".join(reversed(binary))

print(to_binary(12))

# =========================================================
# //* Pro e contro del mio codice
# =========================================================

# //* Pro
# - Soluzione corretta per numeri maggiori di 0.
# - Hai applicato correttamente l'algoritmo richiesto.
# - Buona comprensione della divisione intera e del resto.
# - Codice abbastanza leggibile.
# - Non hai utilizzato funzioni già pronte come bin().

# //! Contro
# - Non gestisce il caso decimal = 0.
# - insert(0, ...) è poco efficiente perché ad ogni inserimento
#   tutti gli elementi della lista vengono spostati.
# - L'if può essere semplificato usando direttamente il resto.
# - I ";" finali non sono necessari in Python.

# =========================================================
# //* Consigli
# =========================================================

# - Ricorda sempre di controllare i casi limite richiesti.
# - Quando costruisci una lista, preferisci append()
#   e inverti il risultato alla fine.
# - Puoi convertire direttamente il resto in stringa:
#
#   binary.append(str(decimal % 2))
#
#   evitando l'if/else.
# - Continua a implementare gli algoritmi manualmente:
#   è un ottimo allenamento per i colloqui.

# =========================================================
# //* Valutazione
# =========================================================

# Livello: Junior+
#
# Complessità temporale: O(n²)
# a causa di insert(0, ...)
#
# Complessità spaziale: O(n)
#
# Passerebbe un colloquio Junior?
# Sì.
#
# Cosa apprezzerebbe un recruiter?
# - Hai seguito esattamente l'algoritmo richiesto.
# - Hai dimostrato di capire come funziona la conversione
#   decimale -> binario.
#
# Caso limite non gestito:
# - to_binary(0) restituisce "" invece di "0".

